From 26e29b776361e73f148003144b988930cc54d3fa Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:26:45 -0400
Subject: [kernel] taskstats: use genl_register_family_with_ops()

Message-id: <529211b9a075ccf239fc756c1b4c8388583d2369.1410528958.git.jbenc@redhat.com>
Patchwork-id: 93953
O-Subject: [RHEL7.1 net PATCH 03/23] taskstats: use genl_register_family_with_ops()
Bugzilla: 1140656
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1140656
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit 88d36a9949513419de3a506e7fca8b82d1dc972a
Author: Johannes Berg <johannes.berg@intel.com>
Date:   Thu Nov 14 17:14:39 2013 +0100

    taskstats: use genl_register_family_with_ops()

    This simplifies the code since there's no longer a
    need to have error handling in the registration.

    Signed-off-by: Johannes Berg <johannes.berg@intel.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/taskstats.c b/kernel/taskstats.c
index 145bb4d..778bb99 100644
--- a/kernel/taskstats.c
+++ b/kernel/taskstats.c
@@ -667,17 +667,18 @@ err:
  nlmsg_free(rep_skb);
 }
 
-static struct genl_ops taskstats_ops = {
- .cmd  = TASKSTATS_CMD_GET,
- .doit  = taskstats_user_cmd,
- .policy  = taskstats_cmd_get_policy,
- .flags  = GENL_ADMIN_PERM,
-};
-
-static struct genl_ops cgroupstats_ops = {
- .cmd  = CGROUPSTATS_CMD_GET,
- .doit  = cgroupstats_user_cmd,
- .policy  = cgroupstats_cmd_get_policy,
+static struct genl_ops taskstats_ops[] = {
+ {
+  .cmd  = TASKSTATS_CMD_GET,
+  .doit  = taskstats_user_cmd,
+  .policy  = taskstats_cmd_get_policy,
+  .flags  = GENL_ADMIN_PERM,
+ },
+ {
+  .cmd  = CGROUPSTATS_CMD_GET,
+  .doit  = cgroupstats_user_cmd,
+  .policy  = cgroupstats_cmd_get_policy,
+ },
 };
 
 /* Needed early in initialization */
@@ -696,26 +697,14 @@ static int __init taskstats_init(void)
 {
  int rc;
 
- rc = genl_register_family(&family);
+ rc = genl_register_family_with_ops(&family, taskstats_ops,
+        ARRAY_SIZE(taskstats_ops));
  if (rc)
   return rc;
 
- rc = genl_register_ops(&family, &taskstats_ops);
- if (rc < 0)
-  goto err;
-
- rc = genl_register_ops(&family, &cgroupstats_ops);
- if (rc < 0)
-  goto err_cgroup_ops;
-
  family_registered = 1;
  pr_info("registered taskstats version %d\n", TASKSTATS_GENL_VERSION);
  return 0;
-err_cgroup_ops:
- genl_unregister_ops(&family, &taskstats_ops);
-err:
- genl_unregister_family(&family);
- return rc;
 }
 
 /*
-- 
1.7.1